# Задание 1

Вернемся к набору данных о видеоиграх.

Ответьте на следующие вопросы:

    1) Как критики относятся к спортивным играм?
    2) Критикам нравятся больше игры на PC или на PS4?
    3) Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

- сформулируйте нулевую и альтернативную гипотезы;
- выберите пороговый уровень статистической значимости;
- опишите полученные результаты статистического теста.

In [1]:
import pandas as pd
from scipy import stats as st

In [2]:
data = pd.read_csv('video_games_sales.csv')
data.describe()

Rank  VGChartz_Score  Critic_Score  User_Score  Total_Shipped  \
count  55792.000000             0.0   6536.000000  335.000000    1827.000000   
mean   27896.500000             NaN      7.213709    8.253433       1.887258   
std    16105.907446             NaN      1.454079    1.401489       4.195693   
min        1.000000             NaN      1.000000    2.000000       0.030000   
25%    13948.750000             NaN      6.400000    7.800000       0.200000   
50%    27896.500000             NaN      7.500000    8.500000       0.590000   
75%    41844.250000             NaN      8.300000    9.100000       1.800000   
max    55792.000000             NaN     10.000000   10.000000      82.860000   

       Global_Sales      NA_Sales     PAL_Sales     JP_Sales   Other_Sales  \
count  19415.000000  12964.000000  13189.000000  7043.000000  15522.000000   
mean       0.365503      0.275541      0.155263     0.110402      0.044719   
std        0.833022      0.512809      0.399257     0.184673      0.129554   
min        0.000000      0.000000      0.000000     0.000000      0.000000   
25%        0.030000      0.050000      0.010000     0.020000      0.000000   
50%        0.120000      0.120000      0.040000     0.050000      0.010000   
75%        0.360000      0.290000      0.140000     0.120000      0.040000   
max       20.320000      9.760000      9.850000     2.690000      3.120000   

               Year   status  Vgchartzscore  
count  54813.000000  55792.0     799.000000  
mean    2005.659095      1.0       7.425907  
std        8.355585      0.0       1.384226  
min     1970.000000      1.0       2.600000  
25%     2000.000000      1.0       6.800000  
50%     2008.000000      1.0       7.800000  
75%     2011.000000      1.0       8.500000  
max     2020.000000      1.0       9.600000

### 1. Как критики относятся к спортивным играм? 

- H0: средняя удовлетворенность критиков спортивными играми <= 7 баллов из 10
- H1: средняя удовлетворенность критиков спортивными играми составляет более 7 баллов из 10 

За alpha будем принимать 0.05

Сравниваем значение по выборке с константой, поэтому нужен One Sample T-Test.

In [3]:
alpha = 0.05
data_score = data[data.Genre == 'Sports'].Critic_Score.dropna()
stat, pvalue = st.ttest_1samp(data_score, 7, alternative='greater')

print(data[data.Genre == 'Sports'].Critic_Score.mean())
print('Statistics=%.3f, pvalue=%.3f' % (stat, pvalue))

if pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, среднее больше 7 (критикам нравятся спортивные игры)')
else:
    print('Не отвергаем нулевую гипотезу, критики не высоко оценивают спортивные игры')

7.291424418604655
Statistics=5.197, pvalue=0.000
Отвергаем нулевую гипотезу, среднее больше 7 (критикам нравятся спортивные игры)


Вывод: в целом, критики достаточно высоко оценивают спортивные игры (выше 7 баллов из 10).

### 2. Критикам нравятся больше игры на PC или на PS4?

- H0: средняя удовлетворенность критиков играми на PC и на PS4 одинаковая  
- H1: средняя удовлетворенность критиков играми на PC и на PS4 различается

За alpha будем принимать 0.05

Сравниваем две независимые выборки, нужен двухвыборочный двухсторонний критерий, будем использовать Welch's Test.

In [4]:
alpha = 0.05

stat, pvalue = st.ttest_ind(data[data.Platform == 'PC'].Critic_Score, data[data.Platform == 'PS4'].Critic_Score, nan_policy='omit', equal_var=False)

print('Средняя оценка критиков для платформы PC:', round(data[data.Platform == 'PC'].Critic_Score.mean(), 2))
print('Средняя оценка критиков для платформы PS4:', round(data[data.Platform == 'PS4'].Critic_Score.mean(), 2))
print('Statistics=%.3f, pvalue=%.3f' % (stat, pvalue))

if pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, удовлетворенность критиков играми на PC и на PS4 различается')
else:
    print('Не отвергаем нулевую гипотезу, критики одиково относятся к играм на PC и PS4')

Средняя оценка критиков для платформы PC: 7.54
Средняя оценка критиков для платформы PS4: 7.9
Statistics=-2.739, pvalue=0.007
Отвергаем нулевую гипотезу, удовлетворенность критиков играми на PC и на PS4 различается


Вывод: разница между средними оценками платформ статистически значима, следовательно, критики выше оценивают игры на PS4, чем на PC.

### 3. Критикам больше нравятся стрелялки или стратегии?

- H0: средняя удовлетворенность критиков стрелялками и стратегиями одинаковая
- H1: средняя удовлетворенность критиков стрелялками и стратегиями различается

За alpha будем принимать 0.05

Сравниваем две независимые выборки, нужен двухвыборочный двухсторонний критерий, будем использовать Welch's Test.

In [5]:
alpha = 0.05

stat, pvalue = st.ttest_ind(data[data.Genre == 'Shooter'].Critic_Score, data[data.Genre == 'Strategy'].Critic_Score, nan_policy='omit', equal_var=False)

print('Средняя оценка критиков для Shooter:', round(data[data.Genre == 'Shooter'].Critic_Score.mean(), 2))
print('Средняя оценка критиков для Strategy:', round(data[data.Genre == 'Strategy'].Critic_Score.mean(), 2))
print('Statistics=%.3f, pvalue=%.3f' % (stat, pvalue))

if pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, удовлетворенность критиков стрелялками или стратегиями различается')
else:
    print('Не отвергаем нулевую гипотезу, критики одиково относятся к стрелялкам и стратегиям')

Средняя оценка критиков для Shooter: 7.29
Средняя оценка критиков для Strategy: 7.43
Statistics=-1.607, pvalue=0.108
Не отвергаем нулевую гипотезу, критики одиково относятся к стрелялкам и стратегиям


Вывод: разница между средними оценками игр статистически не значима, следовательно, в целом, критики одиково оценивают игры-стрелялки и игры-стратегии.

# Задание 2

Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений по признаку спама. Для этого:
1) Приведите весь текст к нижнему регистру;<br>
2) Удалите мусорные символы;<br>
3) Удалите стоп-слова;<br>
4) Приведите все слова к нормальной форме;<br>
5) Преобразуйте все сообщения в вектора TF-IDF.
Можете поэкспериментировать с параметрами TfidfVectorizer;<br>
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;<br>
7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;<br>
8) Опишите результаты при помощи confusion_matrix;<br>
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [6]:
import re

In [7]:
data_spam = pd.read_csv('spam.csv')
data_spam.Message[5] # здесь и далее в качестве проверки

"FreeMsg Hey there darling it's been 3 week's now and no word back! I'd like some fun you up for it still? Tb ok! XxX std chgs to send, £1.50 to rcv"

#### 1. Приведем  весь текст к нижнему регистру

In [8]:
messages_lower_case = [mess.lower() for mess in data_spam.Message]
messages_lower_case[5] # проверка

"freemsg hey there darling it's been 3 week's now and no word back! i'd like some fun you up for it still? tb ok! xxx std chgs to send, £1.50 to rcv"

#### 2. Удалим мусорные (служебые) символы 

In [9]:
messages_words = [re.sub('[\W_\d]+',' ', mess) for mess in messages_lower_case]
messages_words[5] # проверка

'freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send to rcv'

#### 3. Удалим стоп-слова

In [10]:
# сначала разделим все на слова
messages_splitted = [mess.split() for mess in messages_words]
len(messages_splitted[5]) # проверка

33

In [11]:
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))
mess_without_sw = [[word for word in mess if word not in stopwords_set] for mess in messages_splitted]
len(mess_without_sw[5]) # проверка

16

#### 4. Приведем все слова к нормальной форме

In [12]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
lemmatized_mess = [[wordnet_lemmatizer.lemmatize(word) for word in mess] for mess in mess_without_sw]
lemmatized_mess[5] # проверка

['freemsg',
 'hey',
 'darling',
 'week',
 'word',
 'back',
 'like',
 'fun',
 'still',
 'tb',
 'ok',
 'xxx',
 'std',
 'chgs',
 'send',
 'rcv']

In [13]:
lemmatized_mess_joined = [' '.join(mess) for mess in lemmatized_mess]
lemmatized_mess_joined[5] # проверка

'freemsg hey darling week word back like fun still tb ok xxx std chgs send rcv'

#### 5. Преобразуем все сообщения в вектора TF-IDF

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(lemmatized_mess_joined)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)

#### 6. Разделяем данные на тестовые и тренировочные в соотношении 30/70

In [15]:
# сначала определим в исходном датасете class_message
data_changed = data_spam.copy()

def add_class_message(row):
    if row.Category == 'ham':
        return 0
    else:
        return 1
    
data_changed['class_message'] = data_changed.apply(add_class_message, axis=1)

In [16]:
# соединим матрицу со столбцом спам / не спам, используя столбец class_message в data_changed
data_concat = pd.concat([data_changed['class_message'], tfidf_matrix], axis=1, sort=False)
data_concat.head()

class_message   aa  aah  aaniye  aaooooright  aathi   ab  abbey  abdomen  \
0              0  0.0  0.0     0.0          0.0    0.0  0.0    0.0      0.0   
1              0  0.0  0.0     0.0          0.0    0.0  0.0    0.0      0.0   
2              1  0.0  0.0     0.0          0.0    0.0  0.0    0.0      0.0   
3              0  0.0  0.0     0.0          0.0    0.0  0.0    0.0      0.0   
4              0  0.0  0.0     0.0          0.0    0.0  0.0    0.0      0.0   

   abeg  ...   zf  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada   èn  〨ud  
0   0.0  ...  0.0    0.0     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0  
1   0.0  ...  0.0    0.0     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0  
2   0.0  ...  0.0    0.0     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0  
3   0.0  ...  0.0    0.0     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0  
4   0.0  ...  0.0    0.0     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0  

[5 rows x 7100 columns]

In [17]:
# разделяем данные на выборку для обучения и тестирования
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_concat.iloc[:,1:], data_concat.class_message, test_size=0.30, random_state=42)

#### 7. Построим модель логистической регрессии, оценим ее точность на тестовых данных

In [18]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(random_state=42).fit(X_train, y_train)
y_pred = clf.predict(X_test)

# определяем точность прогноза 
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.9599282296650717

#### 8. Опишем результаты при помощи confusion_matrix

In [19]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[1445,    3],
       [  64,  160]])

In [20]:
from sklearn.metrics import classification_report
print('Classification Report : ')
print (classification_report(y_test, y_pred))

Classification Report : 
              precision    recall  f1-score   support

           0       0.96      1.00      0.98      1448
           1       0.98      0.71      0.83       224

    accuracy                           0.96      1672
   macro avg       0.97      0.86      0.90      1672
weighted avg       0.96      0.96      0.96      1672



Вывод: всего было 1445 истинных положительных результатов, 3 ложных положительных результатов, 64 ложных отрицательных и 160 истинных отрицательных.
    Precision (0.98) говорит нам, что в 98% случаев модель распознает спам.
    Recall (0.71) говорит нам, что в 29% случаев нормальное письмо попадет в спам.

#### 9. Построим датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного)

In [27]:
# соотнесем предсказания с индексами исходного датасета, сохраняя при этом только те, что классифицированы неправильно
data_pred = pd.DataFrame(y_test)
data_pred['Predictions'] = pd.Series(y_pred, index=data_pred.index)
data_pred_ = data_pred[data_pred.class_message != data_pred.Predictions]
del data_pred_['class_message']
data_pred_.head()

Predictions
2952            1
881             0
1961            0
3864            0
2575            0

In [29]:
data_final = pd.merge(data_changed, data_pred_, left_index=True, right_index=True)
del data_final['class_message']
data_final = data_final[['Message', 'Category', 'Predictions']]

def return_category(row):
    if row['Predictions'] == 0:
        return 'ham'
    else:
        return 'spam'

data_final['Predictions'] = data_final.apply(return_category, axis=1)
data_final.head(20) # итоговый датафрейм

Message Category Predictions
15   XXXMobileMovieClub: To use your credit, click ...     spam         ham
19   England v Macedonia - dont miss the goals/team...     spam         ham
68   Did you hear about the new "Divorce Barbie"? I...     spam         ham
95   Your free ringtone is waiting to be collected....     spam         ham
135  Want 2 get laid tonight? Want real Dogging loc...     spam         ham
191  Are you unique enough? Find out from 30th Augu...     spam         ham
227  Will u meet ur dream partner soon? Is ur caree...     spam         ham
305  SMS. ac Blind Date 4U!: Rodds1 is 21/m from Ab...     spam         ham
309  TheMob> Check out our newest selection of cont...     spam         ham
333  Call Germany for only 1 pence per minute! Call...     spam         ham
415  100 dating service cal;l 09064012103 box334sk38ch     spam         ham
495                   Are you free now?can i call now?      ham        spam
505                                            #ERROR!     spam         ham
517  Boltblue tones for 150p Reply POLY# or MONO# e...     spam         ham
731  Email AlertFrom: Jeri StewartSize: 2KBSubject:...     spam         ham
751  Do you realize that in about 40 years, we'll h...     spam         ham
752  You have an important customer service announc...     spam         ham
763  Urgent Ur £500 guaranteed award is still uncla...     spam         ham
856  Talk sexy!! Make new friends or fall in love i...     spam         ham
881  Reminder: You have not downloaded the content ...     spam         ham